### DISCUSSION TITLE
Switching between multiple servers  (multiple profiles)

### DISCUSSION THREAD

**Discussion by GraphicHealer** [rustdesk/rustdesk#1258] [Q&A] [answered]
Would it be possible to add an option to save multiple remote servers (ID/Relay IP and Key) and have the option to switch between them? 
I have one sever setup for work, and I use it to do IT helpdesk on work related PCs, and I also have a personal server meant for helping family and friends. I am currently having to use two seperate PCs to do this. Would you be able to add an option to solve this issue?

Thanks,
- G

**Comment by pedjas** [comment]
It seems RustDesk is not meant to be used with multiple servers. You have to choose one.

The only way to use it with multiple hosts is to have portable exe for each host and do not run them at the same time. If you have installed version then portables would not work besides that one.

**Reply by GraphicHealer** [reply]
I know at the moment it is not meant to be used with multiple hosts, what I was asking was if it would be possible to add the capability. It was a feature request, not a bug fix.

**Reply by pedjas** [reply]
Well, I posted that as feature request some time ago and it was instantly rejected. Without any explanation.

**Comment by IanHarker** [comment]
Yes please, that would be a great addition, as i have multiple companies, and they like to use rustdesk, so if i have to connect, i have to re-input their id and key, and then reset back once finished

**Comment by rustdesk** [comment]
Planed in 1.3 https://github.com/rustdesk/rustdesk/discussions/918, you know we are still working on 1.2

**Reply by Trogvars** [reply]
Thanks!

**Reply by putude** [reply]
Great!! Thanks!!

**Reply by jamiegau** [reply]
Any way to deal with this better is welcomed.  I deal with multiple companies, they are all adopting rustdesk as its very good, and having your own private server in many whys is far more secure then the global solutions.  Its also so much faster when we setup a company only server in our region.

**Reply by somethingstarted** [reply]
thank you so much. please allow an additional no-server profile too? like the default profile. it might be easily overlooked to have profiles be added and then magically not be able to use the no profile default.

**Comment by jamiegau** [comment]
On this topic,  not just multiple servers, but the ability to use them all at the same time.
For example.  I may have 2 clients and a general internet user.  I would want to configure rustdesk to try client1-server, then client2-server then global-servers for the client lookup before it gives up.
You will also need the ability to register a client on 1 or more specific private-servers AND the global-server at the same time.  So either method can be used to look up the address-number.

We need the ability to NOT lock rustdesk into a specific server.

I use it personally with my own server, but have many friends I help out who are on the global server, and having to go into the config and add and remote the ID-server all the time is a pain, a huge pain.

I also consider this a major issue that needs to be addressed sooner than later as it specifically hinders its usefullness and growth potential.  I lot of users would pass on rustdesk due to these issues.

**Reply by GraphicHealer** [reply]
That is actually a brilliant idea. I would love it if that works like that.

The problem I can foresee for it though, is each rust desk server assigns a random ID to its user. There are only so many to go around, and it may end up assigning an ID that already exists on another server. If that's the case then you'll only ever get server 1 with that ID, rather than server 2, which maybe what you need. 

I don't know if there would be a way around that or not. Maybe some sort of manual server selection on top of the auto server selection.

**Reply by jamiegau** [reply]
Excellent point.  or only show a number that can be registered on all servers as the same number.  The Global-server may need to get bigger numbers. 10 digitals over 9 as a general rule as if the rustdesk grows in use like it could.  I know TViewer has more digits for an address.  Typically, if you find a free number on global, the private servers should also have them free.  Would be rare otherwise.

Thats how I would deal with this problem.

**Reply by jannoke** [reply]
I think solution would be to have "id@server" connection name for multiserver envs. So you can actually have same id , but server name behind @ would determine where it actually lies. "@server" could be ip, dns name or just a profile name on the client machine who is connecting.

**Reply by pedjas** [reply]
I agree with jannoke. Asking for RustDesk to search through server to find user is rather complicated and unnecessary task. IT should be as simple as user providing his contact information that contains username and server he is connected to, so you can connect to him over that server (providing you do have privilege to do so - you are able to connect to the same server). I would not go further than that.

**Reply by pedjas** [reply]
I do not think that forcing users (clients) to have same ID on multiple sevrers is not proper way to go. Each serer should provide his own ID's for users independently. 

There so no need to force same ID and there is also privacy issue. In European Union such approach may cause software to be banned by law, as that would allow to track user over multiple servers which is big NO-NO in EU. 

That also cause issues with reserving ID numbers over multiple servers. I already met with that problem when I first connected client to local hosted server and got one ID, but when connected to public server later, it could not use same ID as it was taken but someone else. 

So, it had to change ID, and then when getting back to local server it continued using new ID it got form public server so it ended up as new user, making lot of mess.

**Reply by somethingstarted** [reply]
Yes please. I would LOVE to be able to have a top level group that is a server you're connecting to, and then within that server / group are your computers for said network & any VPN or ztna creds you may need for a network that only allows lan RMM 

Then for another company you have a group that is in their server and say they can RMM without a VPN, you set that one up separately. 

Then you have your group for computers that aren't on any server but on the default public rust desk servers 

The perfect world would be able to use all 3 scenarios at the same time. 

Currently you practically need a different VM for everything company you wanna RMM

**Reply by somethingstarted** [reply]
> I think solution would be to have "id@server" connection name for multiserver envs. So you can actually have same id , but server name behind @ would determine where it actually lies. "@server" could be ip, dns name or just a profile name on the client machine who is connecting.

oooh i like that idea. hell, if they do that they might as well add AD / Ldap support too

**Comment by jamiegau** [comment]
On consideration, I feel the ID@domain.com as the connection number is the best path.
That makes it so the ID server configuration is independant from the ID-Address.  You can connect to any ID on any server without a detached configuration of the setting the ID server before trying to connect.   ID numbers would also be unique as they include the @host.name.com.   for example, connect to. 123456789@my.relay.com.   there is no confusion between 123456789 or 123456789@another.relay.com

The only question I have is the RELAY server.  does it default to the same ID-Server, or does it use the Relay server as specified in the options?

**Reply by pedjas** [reply]
It was nicely done in XMPP/Jabber protocol. User identifies with username and domain, but domain is not server address, just identification of server.

Client asks DNS for special DNS SRV records for specified domain to get server information. In case of RustDesk, that means DNS can provide info of addresses for rendezvous and relay servers, ports and so on.

See https://prosody.im/doc/dns

**Reply by ForsakenRei** [reply]
I felt this is the best way, like you are connecting ssh with user@host, here we will have client_id@name_server.

**Comment by liubq** [comment]
我有一个想法，同时也在做这件事情，Rustdesk不是有API服务吗，我考虑在API服务上增加一个管理服务地址的，可以将所有的服务、中继服务都保存下来，客户端通过选择选择服务器来决定用哪一个服务器，首次选一个服务器，生成ID，并记录到API中，这样，切换服务器时同时J将这个服务器的ID，保存到Rustdesk 客户端中，并应用这台服务器

**Reply by liubq** [reply]
This is from the translation software：
I have an idea, and at the same time, I am also doing this thing, doesn't Rustdesk have API service? I consider adding a management service address to the API service, which can save all the services and relay services. The client decides which server to use by selecting a server, selecting a server for the first time and generating an ID. In this way, J will save the ID of this server to the Rustdesk client when switching servers, and apply this server

**Comment by rustdesk** [comment, marked as answer]
Below works on >=1.2.4. https://github.com/rustdesk/rustdesk/releases/tag/nightly
![image](https://github.com/rustdesk/rustdesk/assets/71636191/cd5e2959-7481-4309-9f1d-54b3027fab54)

> https://github.com/rustdesk/rustdesk/issues/6198#issuecomment-1794694273

**Reply by W-i-n-7** [reply]
Can this be made into a quick dropdown menu to which you can add servers in settings?
+ Per favorite server selection

**Reply by radiocooke** [reply]
> Below works on >=1.2.4. 
> 
> > [#6198 (comment)](https://github.com/rustdesk/rustdesk/issues/6198#issuecomment-1794694273)

I noticed that when using this method, I am able to connect to the client without issue, but the status within the RustDesk app always shows offline. Is this a known condition or did I do something incorrectly?

**Reply by biomassboilers** [reply]
I tried this with a test self hosted server, but always got "port unavailable" errors in Rustdesk v1.4.5. I tested with ports 21115-21119, no luck. The self hosted server works if I put it in Settings >> Network >> ID/Relay server, but does not work when accessed via `1234567890@myserver.local:21117?key=foobar`, nor with any variations of port or IP.

**Reply by Joaquim3** [reply]
1. Have you opened the ports in your Server Firewall ?
2. If it's a cloud server, you certainly must also open the same ports on your Server provider Interface. 

21114:21119/tcp            ALLOW IN    Anywhere                  
8000/tcp                   ALLOW IN    Anywhere                  
21116/udp                  ALLOW IN    Anywhere                  
21114:21119/tcp (v6)       ALLOW IN    Anywhere (v6)             
8000/tcp (v6)              ALLOW IN    Anywhere (v6)             
21116/udp (v6)             ALLOW IN    Anywhere (v6)

**Reply by SebSeifert** [reply]
> > Below works on >=1.2.4. 
> > 
> > > [#6198 (comment)](https://github.com/rustdesk/rustdesk/issues/6198#issuecomment-1794694273)
> 
> I noticed that when using this method, I am able to connect to the client without issue, but the status within the RustDesk app always shows offline. Is this a known condition or did I do something incorrectly?

Same here. Maybe this could be fixed?

**Reply by KoeSource** [reply]
`1234567890@myserver.local:21117?key=foobar`

This works locally.
But it doesn't work via Dyndns from another network.

`Connection via the relay server failed....`

I changed the ports for hbbs and hbbr.
hbbs: 99991
hbbr: 88881

I made the following changes:
```
systemctl edit rustdesksignal.service
[Service]
ExecStart=
ExecStart=/opt/rustdesk/hbbs -p 99991
```

```
systemctl edit rustdeskrelay.service
[Service]
ExecStart=
ExecStart=/opt/rustdesk/hbbr -p 88881
```

I also forwarded the ports on the router’s firewall.
tcp/udp 99990-99999
tcp/udp 88880-88889
tcp/udp 8000

What did I miss?
Why isn't this working?

**Comment by Joaquim3** [comment]
I did this SH workaround, but it's a bit dirty. But it's working fine. Hope this can help.
Check the attached TXT file.
[rustdesk.txt](https://github.com/user-attachments/files/15935726/rustdesk.txt)

**Reply by xlionjuan** [reply]
Not need to upload files, just paste it here
```bash
#!/bin/bash

#---------------------------------------------------------------
#   -> filename : rustdesk.sh
#	-> sudo chmod u+x rustdesk.sh
#	-> Usage : ./rustdesk.sh
#----------------------------------------------------------------------------------------------------------------

clear
echo "┌────────────────────────────────────────────────────────────────────┐"
echo "│ This app is used for RUSTDESK                                      │"
echo "│ Purpose : restart rustdesk service after having selected a Network │"
echo "└────────────────────────────────────────────────────────────────────┘"
echo
echo "┌─────────────────────────────────────┐"
echo "│ Please, Select your Network below ! │"
echo "└─────────────────────────────────────┘"
echo "(1) - 192.168.1.2"
echo "(2) - xxx.xxx.xxx.xxx"

#rustdeskconfig file
rustdeskfile="/home/USERNAME/.config/rustdesk/RustDesk2.toml"

#asks the user to select his network
read -p "Network : " rustdesknetwork


#replace content of rustdeskconfig file using user's choice
case $rustdesknetwork in

  1) cat > $rustdeskfile << EOF
rendezvous_server = '192.168.1.2:21116'
nat_type = 1
serial = 0
[options]
allow-linux-headless = 'Y'
custom-rendezvous-server = '192.168.1.2'
local-ip-addr = '192.168.1.9'
direct-server = 'Y'
key = 'XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX'
verification-method = 'use-permanent-password'
EOF
    echo "192.168.1.2 Selected."
    ;;


  2) cat > $rustdeskfile << EOF
rendezvous_server = 'xxx.xxx.xxx.xxx:21116'
nat_type = 1
serial = 0
[options]
allow-linux-headless = 'Y'
custom-rendezvous-server = 'xxx.xxx.xxx.xxx'
local-ip-addr = '192.168.1.9'
direct-server = 'Y'
key = 'XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX'
verification-method = 'use-permanent-password'
EOF

    echo "xxx.xxx.xxx.xxx Selected."
    ;;


  *)
    echo "bad choice: $rustdesknetwork"
    ;;
esac

#restart rustdesk service
echo "MYPASSWORD" | sudo -S service rustdesk restart 2> /dev/null

#stop rustdesk application
echo "MYPASSWORD" | sudo -S pkill rustdesk 2> /dev/null
```

**Reply by Joaquim3** [reply]
**New Clean VERSION**
Works fine, 
But only one Network can be displayed at a time. Of course, you can have many devices on each Network.

```shell
#!/bin/bash

  #---------------------------------------------------------------
  #     -> filename : rustdesk.sh
  #	-> sudo chmod u+x rustdesk.sh
  #	-> Usage : ./rustdesk.sh
  #----------------------------------------------------------------------------------------------------------------



    clear

    echo "┌────────────────────────────────────────────────────────────────────────────┐"
    echo "│ This app is used for RUSTDESK                                              │"
    echo "│ Purpose : restart rustdesk service and app after having selected a Network │"
    echo "└────────────────────────────────────────────────────────────────────────────┘"
    echo
    echo "┌─────────────────────────────────────┐"
    echo "│ Please, Select your Network below ! │"
    echo "│─────────────────────────────────────│"
    echo "│(0) - no server                      │"
    echo "│(1) - CLIENT 1 : xxx.xxx.xxx.xxx     │"
    echo "│(2) - CLIENT 2 : zzz.zzz.zzz.zzz     │"
    echo "│(*) - quit                           │"
    echo "└─────────────────────────────────────┘"



    #----------------------------------------------------------
    #this function replaces rustdeskfile and restarts rustdesk
    fnc_restart_rustdesk() {

        array=("$@")
        #---------------------------------
        # saving array to file
        #---------------------------------
        rm $rustdeskfile
        for element in "${array[@]}"
        do
          echo "$element" >> $rustdeskfile
        done

        echo $textdisplay

        #restart rustdesk service
        #trust zone only for clear password !!
        echo "MYPASSWORD" | sudo -S service rustdesk restart 2> /dev/null

        #stop rustdesk application
        echo "MYPASSWORD" | sudo -S killall rustdesk 2> /dev/null
        echo "RustDesk Service Restarted"
        echo
        nohup rustdesk &>/dev/null &
    }




    #----------------------------------------------------------
    #get the local IP Address
    mylocalip=$(hostname  -I | cut -f1 -d' ')

    #rustdeskconfig file
    rustdeskfile="/home/MYUSERNAME/.config/rustdesk/RustDesk2.toml"

    #asks the user to select his network
    read -p "Network : " rustdesknetwork



      #replace content of rustdeskconfig file using user's choice
      case $rustdesknetwork in


        #---------------------------------
        0) #user selected no server

            array=("")
                textdisplay="no server : Selected."
                fnc_restart_rustdesk "${array[@]}"
            ;;



        #---------------------------------
        1) #user selected xxx.xxx.xxx.xxx

            array=("rendezvous_server = 'xxx.xxx.xxx.xxx:21116'"
                  "nat_type = 1"
                  "serial = 0"
                  "[options]"
                  "allow-linux-headless = 'Y'"
                  "custom-rendezvous-server = 'xxx.xxx.xxx.xxx'"
                  "local-ip-addr = '$mylocalip'"
                  "direct-server = 'Y'"
                  "key = 'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx'"
                  "verification-method = 'use-permanent-password'")
                  textdisplay="CLIENT 1 : xxx.xxx.xxx.xxx Selected."
                  fnc_restart_rustdesk "${array[@]}"
            ;;



        #---------------------------------
        2) #user selected zzz.zzz.zzz.zzz

            array=("rendezvous_server = 'zzz.zzz.zzz.zzz:21116'"
                  "nat_type = 1"
                  "serial = 0"
                  "[options]"
                  "allow-linux-headless = 'Y'"
                  "custom-rendezvous-server = 'zzz.zzz.zzz.zzz'"
                  "local-ip-addr = '$mylocalip'"
                  "direct-server = 'Y'"
                  "key = 'zzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzzz'"
                  "verification-method = 'use-permanent-password'")
                  textdisplay="CLIENT 2 : zzz.zzz.zzz.zzz Selected."
                  fnc_restart_rustdesk "${array[@]}"
            ;;



        #---------------------------------
        *) #user selected any other choice

            echo "no choice: bye"
            ;;


      esac


```

**Comment by Accademia** [comment]
> Would it be possible to add an option to save multiple remote servers (ID/Relay IP and Key) and have the option to switch between them? 
> 
> I have one sever setup for work, and I use it to do IT helpdesk on work related PCs, and I also have a personal server meant for helping family and friends. I am currently having to use two seperate PCs to do this. Would you be able to add an option to solve this issue?
> 
> 
> 
> Thanks,
> 
> - G

Specifically, if I set up multiple ID servers and relay servers, I would like to be able to configure them all in my client simultaneously. This way, if one server fails (especially if it is blocked by the Great Firewall of China), the client can automatically connect to another set of servers, thus avoiding disruptions in communication between multiple clients.

Please note, what I am looking for is the ability to configure multiple servers on the client side, not just multiple relay servers on the server side ( https://rustdesk.com/docs/en/self-host/rustdesk-server-pro/relay ). The latter would still result in the inability to log in if the ID server is blocked by the Great Firewall of China.

**Reply by HomeWiseAdmin** [reply]
Please implmenet choosing the network config in the UI or in a config file.   Thanks, Lets upvote this as a feature request

**Comment by HomeWiseAdmin** [comment]
Please implmenet choosing the network config in the UI or in a config file.   Thanks, Lets upvote this as a feature request

**Reply by Joaquim3** [reply]
Yes, It would be nice to have that implemented. 
I have also several servers using RustDesk for various groups.
It is annoying to copy/paste every time the ID and the Key when I have to support them.

**Comment by Transis-Felipe** [comment]
Or just a switch on main window to enable/disable self-hosted

![image](https://github.com/user-attachments/assets/5981aebe-89c1-4790-916d-274047c5da5b)

**Reply by HomeWiseAdmin** [reply]
I am running the latest version 1.4.0, except I do not have that witch/button, see for yourself :-) Am I missing anything ? 

Not to mention, what to do when you have multiple self-hosted servers !!!

![Snapshot-001-0614-133116](https://github.com/user-attachments/assets/f2c3bede-a2f6-4770-8009-63a3c108a9f1)

**Reply by HomeWiseAdmin** [reply]
Oh, I see, Speaking of mockups, here is one.
![RD select server](https://github.com/user-attachments/assets/1299c361-61a7-48a3-ab31-0bfed22ab47b)

**Comment by GraphicHealer** [comment]
> Or just a switch on main window to enable/disable self-hosted
> 
> ![image](https://github.com/user-attachments/assets/5981aebe-89c1-4790-916d-274047c5da5b)
> 

Except, what if you have multiple self-hosted servers? Like a personal one, and then one you're using for work?

**Reply by Joaquim3** [reply]
If you use my code it works :  **https://github.com/rustdesk/rustdesk/discussions/1258#discussioncomment-10065862**

**With this small code you are already gaining some time, because, you don't have to type the Server IP and the Key each time you change server.**

**You have just 1 click and 1 key pressed !**

On My Limux Mint, I created a shortcut to my "./rustdesk.sh", so when I need, I just click on it and it launches the terminal with the app.

Of course, this is only a work around, because you can not yet interact with computers using multiple Rustdesk servers **at the same time**. 

**But it has been helping me a lot.**

**Comment by myB0B** [comment]
the better choice is to move server id and key configuration per client and remove it from global app settings
